### PR TITLE
Folding/o1VM: use btreemap instead of hashmap + make structs/types pu…

### PR DESCRIPTION
**Pull request by dannywillems**
blic

### CODE REVIEW

File: optimism/src/main.rs [hunk 1]
```diff
@@ -88,17 +98,38 @@ pub fn main() -> ExitCode {
...
+                            .iter()
+                            .map(|x| FoldingCompatibleExpr::from(x.clone()))
+                            .collect(),
+                    )
+                })
+                .collect();
+        DecomposableFoldingScheme::<MIPSFoldingConfig>::new(
+            constraints,
+            vec![],
+            // FIXME
```
Comment (dannywillems) [inline optimism/src/main.rs:126]: The reviewer can request a change here if too critical. We should pass a ref.

Comment (dannywillems) [reply optimism/src/main.rs:126]: Done in 824c99f228512ee65b16886664d883b21b0df931.

File: folding/src/decomposable_folding.rs [hunk 2]
```diff
@@ -14,17 +14,17 @@ use ark_poly::{Evaluations, Radix2EvaluationDomain};
 use poly_commitment::{PolyComm, SRS};
 use std::collections::BTreeMap;
 
-pub struct DecomposableFoldingScheme<CF: FoldingConfig> {
-    inner: FoldingScheme<CF>,
+pub struct DecomposableFoldingScheme<'a, CF: FoldingConfig> {
```
Comment (querolita) [inline folding/src/decomposable_folding.rs:17]: Why the need now?

File: optimism/src/keccak/tests.rs [hunk 3]
```diff
@@ -613,7 +613,7 @@ fn test_keccak_decomposable_folding() {
     let (_scheme, _final_constraint) = DecomposableFoldingScheme::<KeccakConfig>::new(
         constraints,
         vec![],
-        srs.clone(),
+        &srs,
```
Comment (querolita) [inline optimism/src/keccak/tests.rs:616]: Oh, to avoid this type of clones maybe?

**Review by querolita — APPROVED** [review]

File: folding/src/decomposable_folding.rs (hunk 2 above)
Comment (dannywillems) [reply folding/src/decomposable_folding.rs:17]: Because of the lifetime of the SRS.

File: optimism/src/keccak/tests.rs (hunk 3 above)
Comment (dannywillems) [reply optimism/src/keccak/tests.rs:616]: Yes, correct.

**Comment by dannywillems** [comment]
Conflicts solved.

File: optimism/src/main.rs [hunk 4]
```diff
@@ -88,17 +98,37 @@ pub fn main() -> ExitCode {
...
-        lookups: Vec::new(),
-    };
+    let mut mips_con_env = mips_constraints::Env::<Fp>::default();
     // The keccak environment is extracted inside the loop
 
     // Initialize the circuits. Includes pre-folding witnesses.
     let mut mips_trace = MIPSTrace::<Fp>::new(DOMAIN_SIZE, &mut mips_con_env);
     let mut keccak_trace = KeccakTrace::<Fp>::new(DOMAIN_SIZE, &mut KeccakEnv::<Fp>::default());
 
+    let _mips_folding = {
```
Comment (dannywillems) [inline optimism/src/main.rs:108]: I can move this in a draft PR the reviewers do not accept this change. This change shows that the MIPS instructions can now be used with the `DecomposableFoldingScheme`, but it is not in a final state.

**Pull request merged**